File: src/ipc/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { ProgressEvent, QueueEvent, SidecarEvent, UpdateProgress } from "@/types";

type Handlers = {
  onProgress?: (e: ProgressEvent) => void;
  onQueue?: (e: QueueEvent) => void;
  onSidecar?: (e: SidecarEvent) => void;
  onUpdateProgress?: (e: UpdateProgress) => void;
};

export async function subscribeAll(h: Handlers): Promise<UnlistenFn> {
  const unlisteners: UnlistenFn[] = [];
  if (h.onProgress) {
    unlisteners.push(
      await listen<ProgressEvent>("goop://queue/progress", (e) => h.onProgress!(e.payload)),
    );
  }
  if (h.onQueue) {
    unlisteners.push(
      await listen<QueueEvent>("goop://queue/state_changed", (e) => h.onQueue!(e.payload)),
    );
  }
  if (h.onSidecar) {
    unlisteners.push(
      await listen<SidecarEvent>("goop://sidecar/yt_dlp_updated", (e) =>
        h.onSidecar!(e.payload),
      ),
    );
    unlisteners.push(
      await listen<SidecarEvent>("goop://sidecar/warning", (e) => h.onSidecar!(e.payload)),
    );
  }
  if (h.onUpdateProgress) {
    unlisteners.push(
      await listen<UpdateProgress>("goop://update/progress", (e) =>
        h.onUpdateProgress!(e.payload),
      ),
    );
  }
  return () => unlisteners.forEach((u) => u());
}

File: src-tauri/src/events.rs
use goop_core::{EventSink, ProgressEvent, QueueEvent, SidecarEvent};
use tauri::{AppHandle, Emitter};

pub struct TauriSink(pub AppHandle);

impl EventSink for TauriSink {
    fn emit_progress(&self, e: ProgressEvent) {
        let _ = self.0.emit("goop://queue/progress", e);
    }
    fn emit_queue(&self, e: QueueEvent) {
        let _ = self.0.emit("goop://queue/state_changed", e);
    }
    fn emit_sidecar(&self, e: SidecarEvent) {
        match &e {
            SidecarEvent::YtDlpUpdated { .. } => {
                let _ = self.0.emit("goop://sidecar/yt_dlp_updated", &e);
            }
            SidecarEvent::Warning { .. } => {
                let _ = self.0.emit("goop://sidecar/warning", &e);
            }
        }
    }
}
